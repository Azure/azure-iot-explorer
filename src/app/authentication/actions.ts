/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import { GET, SET } from '../constants/actionTypes';

const actionCreator = actionCreatorFactory('AUTHENTICATION');
export const setLoginPreferenceAction = actionCreator.async<string, void>(SET);
export const getLoginPreferenceAction = actionCreator.async<void, string>(GET);
