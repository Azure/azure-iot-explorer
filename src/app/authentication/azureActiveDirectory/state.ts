/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { IotHubDescription } from '../../api/models/iotHubDescription';
import { AzureSubscription } from '../../api/models/azureSubscription';
export interface AzureActiveDirectoryStateInterface {
    formState: 'initialized' | 'working' | 'failed' | 'idle' | 'keyPicked';
    iotHubKey: string;
    iotHubs: IotHubDescription[];
    subscriptions: AzureSubscription[];
    token: string;
}

export const getInitialAzureActiveDirectoryState = (): AzureActiveDirectoryStateInterface => ({
    formState: 'initialized',
    iotHubKey: undefined,
    iotHubs: [],
    subscriptions: [],
    token: undefined
});
