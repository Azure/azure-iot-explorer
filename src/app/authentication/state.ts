/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
export enum AuthenticationMethodPreference {
    AzureAD = 'AzureAD',
    ConnectionString = 'ConnectionString'
}

export interface AuthenticationStateInterface {
    formState: 'initialized' | 'working' | 'failed' | 'idle';
    preference: string;
}

export const getInitialAuthenticationState = (): AuthenticationStateInterface => ({
    formState: 'initialized',
    preference: undefined
});
