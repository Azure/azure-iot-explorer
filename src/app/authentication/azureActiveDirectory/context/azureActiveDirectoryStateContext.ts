/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { getInitialAzureActiveDirectoryState, AzureActiveDirectoryStateInterface } from '../state';
import { AzureActiveDirectoryInterface } from './azureActiveDirectoryStateProvider';

export const getInitialAzureActiveDirectoryOps = (): AzureActiveDirectoryInterface => ({
    getIotHubKey: () => undefined,
    getIotHubs: () => undefined,
    getSubscriptions: () => undefined,
    getToken: () => undefined,
    login: () => undefined,
    logout: () => undefined
});

export const AzureActiveDirectoryStateContext = React.createContext<[AzureActiveDirectoryStateInterface, AzureActiveDirectoryInterface]>
    ([
        getInitialAzureActiveDirectoryState(),
        getInitialAzureActiveDirectoryOps()
    ]);
export const useAzureActiveDirectoryStateContext = () => React.useContext(AzureActiveDirectoryStateContext);
