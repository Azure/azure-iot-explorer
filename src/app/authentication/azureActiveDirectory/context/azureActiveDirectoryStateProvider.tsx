import * as React from 'react';
import { getInitialAzureActiveDirectoryState } from '../state';
import { azureActiveDirectoryReducer } from '../reducers';
import { useAsyncSagaReducer } from '../../../shared/hooks/useAsyncSagaReducer';
import { AzureActiveDirectoryStateContext } from './azureActiveDirectoryStateContext';
import { getUserProfileTokenAction, getSubscriptionListAction, loginAction, logoutAction, getIotHubsBySubscriptionAction, getIoTHubKeyAction } from '../actions';
import { azureActiveDirectorySaga } from '../saga';

export interface AzureActiveDirectoryInterface {
    getIotHubKey(hubId: string, hubName: string): void;
    getIotHubs(subscriptionId: string): void;
    getSubscriptions(): void;
    getToken(): void;
    login(): void;
    logout(): void;
}

export const AzureActiveDirectoryStateContextProvider: React.FC = props => {
    const [state, dispatch] = useAsyncSagaReducer(azureActiveDirectoryReducer, azureActiveDirectorySaga, getInitialAzureActiveDirectoryState(), 'azureActiveDirectoryState');

    const azureActiveDirectoryApi: AzureActiveDirectoryInterface = {
        getIotHubKey: (hubId: string, hubName: string) => dispatch(getIoTHubKeyAction.started({hubId, hubName})),
        getIotHubs: (subscriptionId: string) => dispatch(getIotHubsBySubscriptionAction.started(subscriptionId)),
        getSubscriptions: () => dispatch(getSubscriptionListAction.started()),
        getToken: () => dispatch(getUserProfileTokenAction.started()),
        login: () =>  dispatch(loginAction.started()),
        logout: () => dispatch(logoutAction.started())
    };

    return (
        <AzureActiveDirectoryStateContext.Provider value={[state, azureActiveDirectoryApi]}>
            {props.children}
        </AzureActiveDirectoryStateContext.Provider>
    );
};
