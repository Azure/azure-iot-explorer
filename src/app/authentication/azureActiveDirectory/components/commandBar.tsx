/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { CommandBar } from '@fluentui/react';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { useAuthenticationStateContext } from '../../context/authenticationStateContext';
import { NAVIGATE_BACK } from '../../../constants/iconNames';
import { useAzureActiveDirectoryStateContext } from '../context/azureActiveDirectoryStateContext';

export const AzureActiveDirectoryCommandBar: React.FC = () => {
    const { t } = useTranslation();
    const [ , { setLoginPreference } ] = useAuthenticationStateContext();
    const [{ token }, { logout, login }] =  useAzureActiveDirectoryStateContext();

    const switchAuth = () => {
        setLoginPreference('');
    };

    const getCommandBarItems = () => {
        const items = [{
            ariaLabel: t(ResourceKeys.authentication.authSelection.switchAuthType),
            iconProps: { iconName: NAVIGATE_BACK },
            key: 'switch',
            onClick: switchAuth,
            text: t(ResourceKeys.authentication.authSelection.switchAuthType)
        }];

        return token ? [{
            ariaLabel: t(ResourceKeys.authentication.azureActiveDirectory.command.logout),
            iconProps: { iconName: 'Signout' },
            key: 'signout',
            onClick: logout,
            text: t(ResourceKeys.authentication.azureActiveDirectory.command.logout)
            }, ...items] :
            [{
                ariaLabel: t(ResourceKeys.authentication.azureActiveDirectory.command.login),
                iconProps: { iconName: 'Signin' },
                key: 'signin',
                onClick: login,
                text: t(ResourceKeys.authentication.azureActiveDirectory.command.login)
            }, ...items];
    };

    return (
        <CommandBar
            items={getCommandBarItems()}
        />
    );
};
