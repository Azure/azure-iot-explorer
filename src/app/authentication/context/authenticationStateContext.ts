/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { getInitialAuthenticationState, AuthenticationStateInterface } from '../state';
import { AuthenticationInterface } from './authenticationStateProvider';

export const getInitialAuthenticationOps = (): AuthenticationInterface => ({
    getLoginPreference: () => undefined,
    setLoginPreference: () => undefined
});

export const AuthenticationStateContext = React.createContext<[AuthenticationStateInterface, AuthenticationInterface]>
    ([
        getInitialAuthenticationState(),
        getInitialAuthenticationOps()
    ]);
export const useAuthenticationStateContext = () => React.useContext(AuthenticationStateContext);
