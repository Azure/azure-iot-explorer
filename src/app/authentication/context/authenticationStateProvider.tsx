/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { getInitialAuthenticationState } from '../state';
import { authenticationReducer } from '../reducers';
import { useAsyncSagaReducer } from '../../shared/hooks/useAsyncSagaReducer';
import { AuthenticationStateContext } from './authenticationStateContext';
import { getLoginPreferenceAction, setLoginPreferenceAction } from '../actions';
import { authenticationSaga } from '../saga';

export interface AuthenticationInterface {
    getLoginPreference(): void;
    setLoginPreference(preference: string): void;
}

export const AuthenticationStateContextProvider: React.FC = props => {
    const [state, dispatch] = useAsyncSagaReducer(authenticationReducer, authenticationSaga, getInitialAuthenticationState(), 'authenticationState');

    const authenticationApi: AuthenticationInterface = {
        getLoginPreference: () => dispatch(getLoginPreferenceAction.started()),
        setLoginPreference: (reference: string) =>  dispatch(setLoginPreferenceAction.started(reference))
    };

    return (
        <AuthenticationStateContext.Provider value={[state, authenticationApi]}>
            {props.children}
        </AuthenticationStateContext.Provider>
    );
};
