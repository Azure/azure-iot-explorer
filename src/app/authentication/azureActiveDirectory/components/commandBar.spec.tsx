/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { shallow } from 'enzyme';
import { CommandBar } from '@fluentui/react';
import { AzureActiveDirectoryCommandBar } from './commandBar';
import { getInitialAzureActiveDirectoryState } from '../state';
import * as azureActiveDirectoryStateContext from '../context/azureActiveDirectoryStateContext';
import * as authenticationStateContext from '../../../authentication/context/authenticationStateContext';
import { getInitialAuthenticationState } from '../../../authentication/state';

describe('AzureActiveDirectoryCommandBar', () => {
    it('matches snapshot when token is not present', () => {
        jest.spyOn(azureActiveDirectoryStateContext, 'useAzureActiveDirectoryStateContext').mockReturnValue(
            [getInitialAzureActiveDirectoryState(), azureActiveDirectoryStateContext.getInitialAzureActiveDirectoryOps()]);
        const wrapper = shallow(<AzureActiveDirectoryCommandBar/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('matches snapshot when token is present', () => {
        jest.spyOn(azureActiveDirectoryStateContext, 'useAzureActiveDirectoryStateContext').mockReturnValue(
            [{...getInitialAzureActiveDirectoryState(), token: 'token'}, azureActiveDirectoryStateContext.getInitialAzureActiveDirectoryOps()]);
        const wrapper = shallow(<AzureActiveDirectoryCommandBar/>);
        expect(wrapper).toMatchSnapshot();
    });


    it('calls api respectively', () => {
        const logout = jest.fn();
        jest.spyOn(azureActiveDirectoryStateContext, 'useAzureActiveDirectoryStateContext').mockReturnValue(
            [{...getInitialAzureActiveDirectoryState(), token: 'token'}, {...azureActiveDirectoryStateContext.getInitialAzureActiveDirectoryOps(), logout}]);

        const setLoginPreference = jest.fn();
        jest.spyOn(authenticationStateContext, 'useAuthenticationStateContext').mockReturnValue(
            [getInitialAuthenticationState(), {...authenticationStateContext.getInitialAuthenticationOps(), setLoginPreference}]);

        const wrapper = shallow(<AzureActiveDirectoryCommandBar/>);

        wrapper.find(CommandBar).props().items[0].onClick(undefined);
        wrapper.update();
        expect(logout).toHaveBeenCalled();

        wrapper.find(CommandBar).props().items[1].onClick(undefined);
        wrapper.update();

        expect(setLoginPreference).toHaveBeenCalledWith('');
    });
});
