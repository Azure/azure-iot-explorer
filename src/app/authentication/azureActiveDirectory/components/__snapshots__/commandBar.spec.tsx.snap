// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AzureActiveDirectoryCommandBar matches snapshot when token is not present 1`] = `
<StyledCommandBarBase
  items={
    Array [
      Object {
        "ariaLabel": "authentication.azureActiveDirectory.command.login",
        "iconProps": Object {
          "iconName": "Signin",
        },
        "key": "signin",
        "onClick": [Function],
        "text": "authentication.azureActiveDirectory.command.login",
      },
      Object {
        "ariaLabel": "authentication.authSelection.switchAuthType",
        "iconProps": Object {
          "iconName": "NavigateBack",
        },
        "key": "switch",
        "onClick": [Function],
        "text": "authentication.authSelection.switchAuthType",
      },
    ]
  }
/>
`;

exports[`AzureActiveDirectoryCommandBar matches snapshot when token is present 1`] = `
<StyledCommandBarBase
  items={
    Array [
      Object {
        "ariaLabel": "authentication.azureActiveDirectory.command.logout",
        "iconProps": Object {
          "iconName": "Signout",
        },
        "key": "signout",
        "onClick": [Function],
        "text": "authentication.azureActiveDirectory.command.logout",
      },
      Object {
        "ariaLabel": "authentication.authSelection.switchAuthType",
        "iconProps": Object {
          "iconName": "NavigateBack",
        },
        "key": "switch",
        "onClick": [Function],
        "text": "authentication.authSelection.switchAuthType",
      },
    ]
  }
/>
`;
